File: src/main/java/icu/takeneko/nfh/progress/Progress.java
package icu.takeneko.nfh.progress;

import net.fabricmc.loader.api.entrypoint.EntrypointContainer;
import org.spongepowered.asm.logging.ILogger;
import org.spongepowered.asm.mixin.transformer.MixinTargetContext;
import org.spongepowered.asm.service.MixinService;

import java.lang.reflect.Field;

public class Progress {
    private static final ILogger logger = MixinService.getService().getLogger("NekoFabricHacks");

    public static void onEntrypointInvoke(String key, EntrypointContainer<?> container) {
        logger.info(
                "Loading {} Entrypoint({}) of Mod {} {}",
                key.substring(0, 1).toUpperCase() + key.substring(1),
                container.getEntrypoint().getClass().getName(),
                container.getProvider().getMetadata().getId(),
                container.getProvider().getMetadata().getVersion()
        );
    }

    public static void onMixinConfigPluginLoad(Object handle) throws Throwable{
        Field f = handle.getClass().getDeclaredField("plugin");
        f.setAccessible(true);
        Object plugin = f.get(handle);
        if (plugin == null)return;
        logger.info("Loading Mixin Config Plugin " + plugin.getClass().getName());
    }

    public static void onSelectingMixinConfig(String name){
        logger.info("Selecting Mixin Config " + name);
    }

    public static void onMixinApply(MixinTargetContext target, String pass){
        logger.info(
                "Applying Mixin [{}(from {}) -> {}] at stage {}",
                target.getMixin().getClassName(),
                target.getMixin().getConfig().getName(),
                target.getClassInfo().getName(),
                pass
        );
    }
}

File: src/main/java/icu/takeneko/nfh/progress/Progress.kt
package icu.takeneko.nfh.progress

import net.fabricmc.loader.api.entrypoint.EntrypointContainer
import org.spongepowered.asm.service.MixinService
import java.lang.reflect.Proxy

object Progress {

    private val logger = MixinService.getService().getLogger("NekoFabricHacks")

    @JvmStatic
    fun onEntrypointInvoke(key: String, container: EntrypointContainer<*>) {
        logger.info(
            "Loading ${describeKey(key)} Entrypoint(${container.entrypoint.javaClass.name}) of Mod ${container.provider.metadata.id} ${container.provider.metadata.version}"
        )
    }

    @JvmStatic
    fun onMixinConfigPluginLoad(handle:Object){
        val plugin = handle.javaClass.getDeclaredField("plugin").apply { isAccessible = true }.get(handle)
        if (plugin == null)return
        logger.info("Loading Mixin Config Plugin ${plugin.javaClass.name}")
    }

    @JvmStatic
    fun onSelectingMixinConfig(name: String){
        logger.info("Selecting Mixin Config $name")
    }

    private fun describeKey(key: String): String {
        return key.replaceFirstChar { it.uppercaseChar() }
    }
}

